zkevm-circuits/src/evm_circuit/execution/error_stack.rs

```diff
@@ -0,0 +1,342 @@
+        // call. Restore caller state to next StepState
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                0.expr(),
+                // rw_offset is handled in construct internally
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+            )
+        });
+
+        Self {
+            opcode,
+            min_stack_pointer,
+            max_stack_pointer,
+            is_overflow,
+            is_underflow,
+            restore_context,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let min_stack = opcode.stack_pair().0 as u64;
+        let max_stack = opcode.stack_pair().1 as u64;
+
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+        // Inputs/Outputs
+        self.min_stack_pointer
+            .assign(region, offset, Value::known(F::from(min_stack)))?;
+        self.max_stack_pointer
+            .assign(region, offset, Value::known(F::from(max_stack)))?;
+
+        self.is_overflow.assign(
+            region,
+            offset,
+            F::from(step.stack_pointer as u64),
+            F::from(min_stack),
+        )?;
+        self.is_underflow.assign(
+            region,
+            offset,
+            F::from(max_stack),
+            F::from(step.stack_pointer as u64),
+        )?;
+
+        self.restore_context
+            .assign(region, offset, block, call, step, 1)?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{test::run_test_circuit, witness::block_convert};
+    use crate::test_util::{run_test_circuits, run_test_circuits_with_params};
+
+    use bus_mapping::circuit_input_builder::CircuitsParams;
+    use bus_mapping::evm::OpcodeId;
+    use eth_types::{
+        self, address, bytecode, bytecode::Bytecode, geth_types::Account, Address, ToWord, Word,
+    };
+
+    use mock::TestContext;
+
+    fn test_stack_underflow(value: Word) {
+        let bytecode = bytecode! {
+            PUSH32(value)
+            POP
+            POP
+            STOP
+        };
+
+        assert_eq!(
+            run_test_circuits(
+                TestContext::<2, 1>::simple_ctx_with_bytecode(bytecode).unwrap(),
+                None
+            ),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn pop_gadget_underflow() {
+        test_stack_underflow(Word::from(0x030201));
+        test_stack_underflow(Word::from(0xab));
+    }
+
+    #[test]
+    fn stack_overflow_simple() {
```

Comment:
`stack_overflow_simple` seems redundant as `test_stack_overflow` is called only at one place.

Reply:
thanks, will check it .

Reply:
added more two calls in `stack_overflow_simple` for different push, updated in `stack_overflow_simple` [fea873d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/fea873dcaa371498fccfda47be88760437797378) thanks!